aten/src/ATen/native/cuda/ForeachPointwiseOp.cu

```diff
@@ -45,36 +86,68 @@ void foreach_pointwise_op_(TensorList input, TensorList tensors1, TensorList ten
-    TORCH_CHECK(input.size() > 0, "Tensor list must have at least one tensor.");                                               \
-    TORCH_CHECK(input.size() ==  tensors1.size(), "Tensor lists must be of the same length.");                                 \
-    TORCH_CHECK(tensors1.size() ==  tensors2.size(), "Tensor lists must be of the same length.");                              \
-                                                                                                                               \
-    if (!can_use_fast_route(input, scalar) ||                                                                                  \
-        !can_use_fast_route(tensors1, tensors2) ||                                                                             \
-        !can_use_fast_route(input, tensors1)) {                                                                                \
-        return at::native::foreach_tensor_##NAME##_slow_(input, tensors1, tensors2, scalar);                                   \
-    }                                                                                                                          \
-                                                                                                                               \
-    foreach_pointwise_op_<OP>(input, tensors1, tensors2, scalar);                                                              \
+#define FOREACH_POINTWISE_OP_SCALAR(NAME, OP)                                                                                         \
+std::vector<Tensor> foreach_tensor_##NAME##_scalar_cuda(TensorList input, TensorList tensors1, TensorList tensors2, Scalar scalar) {  \
+    TORCH_CHECK(input.size() > 0, "Tensor list must have at least one tensor.");                                                      \
+    TORCH_CHECK(input.size() ==  tensors1.size(), "Tensor lists must be of the same length.");                                        \
+    TORCH_CHECK(tensors1.size() ==  tensors2.size(), "Tensor lists must be of the same length.");                                     \
+                                                                                                                                      \
+    if (!can_use_fast_route(input, scalar) ||                                                                                         \
+        !can_use_fast_route(tensors1, tensors2) ||                                                                                    \
+        !can_use_fast_route(input, tensors1)) {                                                                                       \
+        return at::native::foreach_tensor_##NAME##_scalar_slow(input, tensors1, tensors2, scalar);                                    \
+    }                                                                                                                                 \
```

Comment:
There's a lot of repetition here (as well as in ForeachOpsKernels.cpp). Maybe it would be nice to make some helper functions?

`check_nonempty_and_same_length(TensorList, TensorList, TensorList)` could be reusable.
`can_use_fast_route(TensorList, TensorList, TensorList, Scalar)` and `can_use_fast_route(TensorList, TensorList, TensorList, ArrayRef<double>)` also seem useful